design-docs/012-python-namespaces

```diff
@@ -0,0 +1,216 @@
+  - `credentials.gcs_credentials`
+  - `credentials.gcs_from_env_credentials`
+  - `credentials.gcs_refreshable_credentials`
+  - `credentials.gcs_static_credentials`
+  - `credentials.s3_anonymous_credentials`
+  - `credentials.s3_credentials`
+  - `credentials.s3_from_env_credentials`
+  - `credentials.s3_refreshable_credentials`
+  - `credentials.s3_static_credentials`
+- `dask`
+  - `dask.computing_meta`
+  - `dask.store_dask`
+- `distributed`
+  - `distributed.extract_sessions`
+  - `distributed.merge_sessions`
+- `garbage`
+  - `garbage.GCSummary`
+- `repository`
+  - `repository.Repository`
+- `session`
+  - `session.Session`
+  - `session.ForkSession`
+- `storage`
+  - `storage.S3Options`
+  - `storage.Storage`
+  - `storage.StorageConcurrencySettings`
+  - `storage.StorageRetriesSettings`
+  - `storage.StorageSettings`
+  - `storage.AnyObjectStoreConfig`
+  - `storage.azure_storage`
+  - `storage.gcs_storage`
+  - `storage.gcs_store`
+  - `storage.http_store`
+  - `storage.in_memory_storage`
+  - `storage.local_filesystem_storage`
+  - `storage.local_filesystem_store`
+  - `storage.r2_storage`
+  - `storage.s3_storage`
+  - `storage.s3_store`
+  - `storage.tigris_storage`
+- `store`
+  - `store.IcechunkStore`
+- `xarray`
+  - `xarray.to_icechunk`
+
+New namespaces:
+
+- `config`
+  - `config.CachingConfig`
+  - `config.CompressionConfig`
+  - `config.CompressionAlgorithm`
+  - `config.ObjectStoreConfig`
+  - `config.RepositoryConfig`
+- `conflict`
+  - `conflict.BasicConflictSolver`
+  - `conflict.Conflict`
+  - `conflict.ConflictDetector`
+  - `conflict.ConflictSolver`
+  - `conflict.ConflictType`
+  - `conflict.VersionSelection`
+- `exceptions`
+  - `exceptions.ConflictError`
+  - `exceptions.IcechunkError`
+  - `exceptions.RebaseFailedError`
+- `logging`
+  - `logging.initialize_logs`,
+  - `logging.set_logs_filter`,
+- `manifests`
+  - `manifests.ManifestConfig`
+  - `manifests.ManifestFileInfo`
+  - `manifests.ManifestPreloadCondition`
+  - `manifests.ManifestPreloadConfig`
+  - `manifests.ManifestSplitCondition`
+  - `manifests.ManifestSplitDimCondition`
+  - `manifests.ManifestSplittingConfig`
+- `snapshots`
+  - `snapshots.SnapshotInfo`
+  - `snapshots.Diff`
+- `virtual`
+  - `virtual.VirtualChunkContainer`
+  - `virtual.VirtualChunkSpec`
+
+## Outstanding questions
+
+- It's generally a bit confusing that there are multiple ways to make most classes, e.g. `local_filesystem_store` and `Storage.new_local_filesystem`. Do we really need both?
+  - If we only kept `Storage` we could get rid of a _lot_ of functions in favour of methods...
+  - Alternatively should base classes that aren't really meant to be called by the user (e.g. `Storage`) be public API at all?
+- What's the difference between `storage.local_filesystem_storage` and `storage.local_filesystem_store`?
+  - Similarly how can we better disambiguate `storage.s3_storage` and `storage.s3_store`?
+- Is there a similar question for credentials?
```

Comment:
Credentials are currently a mess

Reply:
These deserve their own separate issue. The class hierarchy is complicated so it's not immediately obvious to me how to simplify the API.